### PR TITLE
perf: replace loc with getitem in Dask

### LINKED ISSUE #1349: [perf]: Dask implementation should use regular getitem over loc
### Describe the bug

Loc isn't very well supported in Dask since all the special implementations that pandas has don't work in a distributed environment. I would generally recommend just using getitem if you want to either select columns or apply a filter. Currently, this blocks column projections for some reason and we don't have many tests for loc specific cases

### Steps or code to reproduce the bug

run q1 for dask but call .optimize(fuse=False).pprint() instead of .compute()

### Expected results

The ReadParquetFSSpec expression should be restricted to a subset of the columns

### Actual results

```
SortValuesBlockwise: sort_function=<methodcaller: sort_values> sort_kwargs={'by': ['l_returnflag', 'l_linestatus'], 'ascending': True, 'na_position': 'first', 'ignore_index': False}
  ResetIndex:
    ColumnsSetter: columns=('sum_qty', 'sum_base_price', 'sum_disc_price', 'sum_charge', 'avg_qty', 'avg_price', 'avg_disc', 'count_order')
      Projection: columns=[('l_quantity', 'sum'), ('l_extendedprice', 'sum'), ('disc_price', 'sum'), ('charge', 'sum'), ('l_quantity', 'mean'), ('l_extendedprice', 'mean'), ('l_discount', 'mean'), ('l_returnflag', 'size')]
        DecomposableGroupbyAggregation(TreeReduce): split_every=8
          DecomposableGroupbyAggregation(GroupByChunk): funcs=[('count-l_discount-600ee3bd25d7c4d5c2a5ddb0de7ca968', <function _apply_func_to_column at 0x10e085440>, {'column': 'l_discount', 'func': <methodcaller: count>}), ('count-l_extendedprice-2e03ca098804aa8d414ad2ca4371a45f', <function _apply_func_to_column at 0x10e085440>, {'column': 'l_extendedprice', 'func': <methodcaller: count>}), ('count-l_quantity-cba6b095366123379c694b2440e68176', <function _apply_func_to_column at 0x10e085440>, {'column': 'l_quantity', 'func': <methodcaller: count>}), ('size-l_returnflag-8b574197810b39f175d748dec390afb9', <function _apply_func_to_column at 0x10e085440>, {'column': 'l_returnflag', 'func': <methodcaller: size>}), ('sum-charge-134b51212e1b19c4d40dfb4ad041f759', <function _apply_func_to_column at 0x10e085440>, {'column': 'charge', 'func': <methodcaller: sum>}), ('sum-disc_price-aa841df1c0667abb221bc0323032012d', <function _apply_func_to_column at 0x10e085440>, {'column': 'disc_price', 'func': <methodcaller: sum>}), ('sum-l_discount-5a1d744b08b5b676964f608e5c33ad45', <function _apply_func_to_column at 0x10e085440>, {'column': 'l_discount', 'func': <methodcaller: sum>}), ('sum-l_extendedprice-3fe60ac654450a52e82c43c590eb25f5', <function _apply_func_to_column at 0x10e085440>, {'column': 'l_extendedprice', 'func': <methodcaller: sum>}), ('sum-l_quantity-7305c5b69ff16f9d055728f74ef859b3', <function _apply_func_to_column at 0x10e085440>, {'column': 'l_quantity', 'func': <methodcaller: sum>})] observed=True dropna=False
            Projection: columns=['l_quantity', 'l_extendedprice', 'l_discount', 'l_returnflag', 'l_linestatus', 'disc_price', 'charge']
              Assign: disc_price=, charge=
                Projection: columns=['l_discount', 'l_extendedprice', 'l_linestatus', 'l_quantity', 'l_returnflag']
                  Loc:
                    ReadParquetFSSpec: path='data/lineitem.parquet' kwargs={'dtype_backend': 'pyarrow'}
                    RenameSeries: index='l_shipdate'
                      RenameSeries: index='l_shipdate'
                        LE: right=datetime.datetime(1998, 9, 2, 0, 0)
                          Projection: columns='l_shipdate'
                            ReadParquetFSSpec: path='data/lineitem.parquet' kwargs={'dtype_backend': 'pyarrow'}
                RenameSeries: index='l_extendedprice'
                  Mul:
                    Projection: columns='l_extendedprice'
                      Loc:
                        ReadParquetFSSpec: path='data/lineitem.parquet' kwargs={'dtype_backend': 'pyarrow'}
                        RenameSeries: index='l_shipdate'
                          RenameSeries: index='l_shipdate'
                            LE: right=datetime.datetime(1998, 9, 2, 0, 0)
                              Projection: columns='l_shipdate'
                                ReadParquetFSSpec: path='data/lineitem.parquet' kwargs={'dtype_backend': 'pyarrow'}
                    RenameSeries: index='l_discount'
                      Sub: left=1
                        Projection: columns='l_discount'
                          Loc:
                            ReadParquetFSSpec: path='data/lineitem.parquet' kwargs={'dtype_backend': 'pyarrow'}
                            RenameSeries: index='l_shipdate'
                              RenameSeries: index='l_shipdate'
                                LE: right=datetime.datetime(1998, 9, 2, 0, 0)
                                  Projection: columns='l_shipdate'
                                    ReadParquetFSSpec: path='data/lineitem.parquet' kwargs={'dtype_backend': 'pyarrow'}
                RenameSeries: index='l_extendedprice'
                  Mul:
                    RenameSeries: index='l_extendedprice'
                      Mul:
                        Projection: columns='l_extendedprice'
                          Loc:
                            ReadParquetFSSpec: path='data/lineitem.parquet' kwargs={'dtype_backend': 'pyarrow'}
                            RenameSeries: index='l_shipdate'
                              RenameSeries: index='l_shipdate'
                                LE: right=datetime.datetime(1998, 9, 2, 0, 0)
                                  Projection: columns='l_shipdate'
                                    ReadParquetFSSpec: path='data/lineitem.parquet' kwargs={'dtype_backend': 'pyarrow'}
                        RenameSeries: index='l_discount'
                          Sub: left=1.0
                            Projection: columns='l_discount'
                              Loc:
                                ReadParquetFSSpec: path='data/lineitem.parquet' kwargs={'dtype_backend': 'pyarrow'}
                                RenameSeries: index='l_shipdate'
                                  RenameSeries: index='l_shipdate'
                                    LE: right=datetime.datetime(1998, 9, 2, 0, 0)
                                      Projection: columns='l_shipdate'
                                        ReadParquetFSSpec: path='data/lineitem.parquet' kwargs={'dtype_backend': 'pyarrow'}
                    RenameSeries: index='l_tax'
                      Add: left=1.0
                        Projection: columns='l_tax'
                          Loc:
                            ReadParquetFSSpec: path='data/lineitem.parquet' kwargs={'dtype_backend': 'pyarrow'}
                            RenameSeries: index='l_shipdate'
                              RenameSeries: index='l_shipdate'
                                LE: right=datetime.datetime(1998, 9, 2, 0, 0)
                                  Projection: columns='l_shipdate'
                                    ReadParquetFSSpec: path='data/lineitem.parquet' kwargs={'dtype_backend': 'pyarrow'}
```
### Please run narwhals.show_version() and enter the output below.

```shell
System:
    python: 3.12.7 | packaged by conda-forge | (main, Oct  4 2024, 15:57:01) [Clang 17.0.6 ]
executable: /Users/patrick/mambaforge/envs/narwhals/bin/python3.12
   machine: macOS-14.3.1-arm64-arm-64bit

Python dependencies:
     narwhals: 1.13.3
       pandas: 2.2.3
       polars: 1.12.0
         cudf:
        modin:
      pyarrow: 18.0.0
        numpy: 2.1.3
```

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by raisadz**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [x] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue # 
- Closes #1349

## Checklist

- [x] Code follows style guide (ruff)
- [ ] Tests added 
- [ ] Documented the changes

### CODE REVIEW

File: narwhals/_dask/dataframe.py [hunk 1]
```diff
@@ -258,7 +258,7 @@ def join(
             )
 
             other_native = (
-                other._native_frame.loc[:, right_on]
+                other._native_frame[right_on]
```
Comment (MarcoGorelli) [inline narwhals/_dask/dataframe.py:261]: same

File: narwhals/_dask/dataframe.py [hunk 2]
```diff
@@ -272,12 +272,12 @@ def join(
...
             )
             return self._from_native_frame(
-                df.loc[df[indicator_token] == "left_only"].drop(columns=[indicator_token])
+                df[df[indicator_token] == "left_only"].drop(columns=[indicator_token])
             )
 
         if how == "semi":
             other_native = (
-                other._native_frame.loc[:, right_on]
+                other._native_frame[right_on]
```
Comment (MarcoGorelli) [inline narwhals/_dask/dataframe.py:280]: same

File: narwhals/_dask/dataframe.py [hunk 3]
```diff
@@ -136,7 +136,7 @@ def select(
             )
             return self._from_native_frame(df)
 
-        df = self._native_frame.assign(**new_series).loc[:, list(new_series.keys())]
+        df = self._native_frame.assign(**new_series)[list(new_series.keys())]
```
[the remark this answers is not in the document]
Comment (raisadz) [reply narwhals/_dask/dataframe.py]: Thanks @MarcoGorelli! I addressed this and the comments below.

**Review by MarcoGorelli — APPROVED** [review]
thanks @raisadz !

**Pull request merged**